support telechat2

shunxing12345 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

## Description

This PR introduces **TeleChat2**, a new model designed for [specific use case, e.g., multilingual conversational AI or general-purpose language modeling]. TeleChat2 supports model sizes of [e.g., 3B, 7B, and 35B] and has been pre-trained and fine-tuned on large datasets across English and Chinese. 

## Motivation

Adding TeleChat2 to the Transformers repository allows the community to benefit from its capabilities, leveraging state-of-the-art performance on [specific benchmarks or tasks]. This inclusion aligns with the repository's goal of supporting a diverse range of models and architectures.

## Key Changes

1. Added `TeleChat2Config` for model configuration.
2. Added `TeleChat2Model`, `TeleChat2ForCausalLM`, and corresponding tokenizers.
3. Included detailed documentation and tests to ensure compatibility and coverage.
4. Updated integration points in `AutoModel` and other relevant files.

## Documentation and Tests

- Model architecture, configuration, and usage examples are documented in [relevant files].
- Added unit tests for pretraining and fine-tuning workflows.



## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @ylacombe, @eustlb
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerzr and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @muellerzr
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

Commit:
add telechat2

Signed-off-by: xiangw2 <xiangw2@chinatelecom.cn>

Commit:
Merge branch 'main' of https://github.com/shunxing12345/transformers

Commit:
fix

Signed-off-by: xiangw2 <xiangw2@chinatelecom.cn>

shunxing12345 committed:
Merge branch 'huggingface:main' into main

Commit:
fix

Signed-off-by: xiangw2 <xiangw2@chinatelecom.cn>

Commit:
fix

Commit:
fix

Commit:
fix

Signed-off-by: xiangw2 <xiangw2@chinatelecom.cn>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 commented:
Hi team, 👋

Just following up on this PR. I understand you may have a lot on your plate, but I’d love to hear your thoughts or feedback on this contribution whenever you have time. Please let me know if there’s anything I can do to assist or improve the PR!

Thanks for your time and all the great work you do maintaining this repository! 🙏

Best,

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
Merge branch 'main' into main

LysandreJik commented:
Hey @shunxing12345 👋 

Most of us are on holidays this week for the new year. The rhythm will pickup early next week and the team will review your PR as soon as possible. Thanks a lot for your patience and your contribution 🤗

LysandreJik commented:
cc @ArthurZucker feel free to ping someone for a first review!

shunxing12345 commented:
> Hey @shunxing12345 👋
> 
> Most of us are on holidays this week for the new year. The rhythm will pickup early next week and the team will review your PR as soon as possible. Thanks a lot for your patience and your contribution 🤗

Thank you for the update and for letting me know! I completely understand that it's a holiday period. I'll patiently wait for the team to review the PR, and I appreciate all your efforts. Wishing you a Happy New Year! 🎉

shunxing12345 commented:
Hi @LysandreJik @,
Hope you had a wonderful holiday! I just wanted to politely follow up on PR #35415. Please let me know if there’s anything I can update or clarify to help move things forward. I appreciate your time and look forward to any feedback. Thank you, and happy new year!

ArthurZucker commented:
Hey! 🤗 Happy new year ! @Cyrilvallez will review this! 🤗

shunxing12345 committed:
Merge branch 'main' into fix_conflicts

shunxing12345 committed:
fix_conflicts

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix conflicts

Signed-off-by: shunxing12345 <747777271@qq.com>

Cyrilvallez commented on src/transformers/models/telechat2/__init__.py:
```diff
@@ -0,0 +1,65 @@
+
+try:
+    if not is_torch_available():
+        raise OptionalDependencyNotAvailable()
+except OptionalDependencyNotAvailable:
+    pass
+else:
+    _import_structure["modeling_telechat2"] = [
+        "TeleChat2ForCausalLM",
+        "TeleChat2ForQuestionAnswering",
+        "TeleChat2ForSequenceClassification",
+        "TeleChat2ForTokenClassification",
+        "TeleChat2Model",
+        "TeleChat2PreTrainedModel",
+    ]
+
+
+if TYPE_CHECKING:
+    from .configuration_telechat2 import TeleChat2Config
+
+    try:
+        if not is_torch_available():
+            raise OptionalDependencyNotAvailable()
+    except OptionalDependencyNotAvailable:
+        pass
+    else:
+        from .modeling_telechat2 import (
+            TeleChat2ForCausalLM,
+            TeleChat2ForQuestionAnswering,
+            TeleChat2ForSequenceClassification,
+            TeleChat2ForTokenClassification,
+            TeleChat2Model,
+            TeleChat2PreTrainedModel,
+        )
+
+
+else:
+    import sys
+
+    sys.modules[__name__] = _LazyModule(__name__, globals()["__file__"], _import_structure, module_spec=__spec__)
```
Let's use our new simplified __init__ standard here instead, you can check out the __init__ for Llama model for example, it should be fairly easy to replicate 🤗

Cyrilvallez commented on src/transformers/models/telechat2/modeling_telechat2.py:
```diff
@@ -0,0 +1,1163 @@
+_CHECKPOINT_FOR_DOC = "TeleAI/TeleChat2-3B"
+_CONFIG_FOR_DOC = "TeleChat2Config"
+
+
+# Copied from transformers.models.llama.modeling_llama.LlamaRMSNorm with Llama->TeleChat2
```
We no longer use the "Copied from"!

Cyrilvallez commented on src/transformers/convert_slow_tokenizer.py:
```diff
@@ -1603,6 +1603,7 @@ def converted(self) -> Tokenizer:
     "CodeLlamaTokenizer": LlamaConverter,
     "GemmaTokenizer": GemmaConverter,
     "Phi3Tokenizer": LlamaConverter,
+    "TeleChat2Tokenizer": LlamaConverter,
```
You should not need this!

Cyrilvallez commented on src/transformers/convert_slow_tokenizers_checkpoints_to_fast.py:
```diff
@@ -30,7 +30,9 @@
 TOKENIZER_CLASSES = {
     # Phi3 uses Llama tokenizer
-    name: getattr(transformers, "LlamaTokenizerFast" if name == "Phi3Tokenizer" else name + "Fast")
+    name: getattr(
+        transformers, "LlamaTokenizerFast" if name in ["Phi3Tokenizer", "TeleChat2Tokenizer"] else name + "Fast"
+    )
```
nor this!

Cyrilvallez reviewed (commented):
Hey! Thanks for the contribution and your patience! From what I can see, this is already in a very healthy state! 🤗 You are on page with our latest Attention standards 👌 However, you miss a modular file! We no longer use the "Copied from" syntax, in favor of a "modular file". You can get started with [this link](https://huggingface.co/docs/transformers/main/en/modular_transformers), and check other models' modular file (look for `modular_xxx.py` in the library, e.g. `modular_mistral`) for examples! We basically use inheritance in the modular, which is then automatically unravelled in the "modeling file" based on the inherited model. So for example, classes which are fully "Copied from" can very simply inherit from the base class 😉
Let me know if you need more guidance/have any questions!

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 commented:
@Cyrilvallez  Thank you for your support! I noticed that some classes or methods differ only in terms of their parameters, while the overall logic remains consistent.

To avoid duplicating code, could you suggest a best practice for handling such cases? For example:

1. Would it be appropriate to introduce a base class or utility method to centralize the shared logic and allow parameter customization through inheritance or function arguments?
2. Alternatively, is there a recommended pattern in the modular design to efficiently handle such cases?

I truly appreciate your insights and guidance on this matter. Thank you for your time and support!

stevhliu commented on docs/source/en/model_doc/telechat2.md:
```diff
@@ -0,0 +1,85 @@
+
+The TeleChat2 model was proposed in [TELECHAT TECHNICAL REPORT](https://arxiv.org/pdf/2401.03804) by TeleAI.
+
+### Summary
+
+The abstract from the paper is the following:
+
+TeleChat is a series of large language models, offering decoder-based language models in various sizes (3B, 7B, and 12B). For each size, we provide both the base pretrained model and the fine-tuned chat model aligned with human preferences. TeleChat leverages a Transformer architecture with features such as SwiGLU activation, advanced attention mechanisms (QKV bias, group query attention), and su ...
```
```suggestion
The abstract from the paper is the following:

*TeleChat is a series of large language models, offering decoder-based language models in various sizes (3B, 7B, and 12B). For each size, we provide both the base pretrained model and the fine-tuned chat model aligned with human preferences. TeleChat leverages a Transformer architecture with features such as SwiGLU activation, advanced attention mechanisms (QKV bias, group query attention), and support for sliding window attention. The models are optimized for bilingual proficiency (English and Chinese) and include an enhanced tokenizer adaptable to diverse natural languages and coding formats.*
```

stevhliu commented on docs/source/en/model_doc/telechat2.md:
```diff
@@ -0,0 +1,85 @@
+
+TeleChat is a series of large language models, offering decoder-based language models in various sizes (3B, 7B, and 12B). For each size, we provide both the base pretrained model and the fine-tuned chat model aligned with human preferences. TeleChat leverages a Transformer architecture with features such as SwiGLU activation, advanced attention mechanisms (QKV bias, group query attention), and su ...
+
+## Usage tips
+
+The original code for telechat2 can be found [here](https://huggingface.co/Tele-AI/TeleChat2-7B).
+
+In the following, we demonstrate how to use `TeleChat2-7B` for the inference. Note that we have used the ChatML format for dialog, in this demo we show how to leverage `apply_chat_template` for this purpose.
```
```suggestion
The original code for telechat2 can be found [here](https://huggingface.co/Tele-AI/TeleChat2-7B).

## Tips

In the following, we demonstrate how to use `TeleChat2-7B` for inference. The example below shows how to use `apply_chat_template` with the ChatML format for dialog.
```

stevhliu reviewed (commented):
Thanks for the docs!

shunxing12345 committed:
change docs

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
Merge branch 'fix_conflicts' of https://github.com/shunxing12345/transformers into fix_conflicts

shunxing12345 committed:
Merge branch 'main' into fix_conflicts

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
Merge branch 'fix_conflicts' of https://github.com/shunxing12345/transformers into fix_conflicts

shunxing12345 commented:
> Thanks for the docs!

@stevhliu  Thank you for your review! I have updated the docs as requested. Please let me know if there's anything else that needs adjustment. 😊

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 commented:
> Hey! Thanks for the contribution and your patience! From what I can see, this is already in a very healthy state! 🤗 You are on page with our latest Attention standards 👌 However, you miss a modular file! We no longer use the "Copied from" syntax, in favor of a "modular file". You can get started with [this link](https://huggingface.co/docs/transformers/main/en/modular_transformers), and check other models' modular file (look for `modular_xxx.py` in the library, e.g. `modular_mistral`) for examples! We basically use inheritance in the modular, which is then automatically unravelled in the "modeling file" based on the inherited model. So for example, classes which are fully "Copied from" can very simply inherit from the base class 😉 Let me know if you need more guidance/have any questions!

Hi @Cyrilvallez ,

Thank you so much for your feedback and detailed guidance! 😊 I’ve addressed the issue and added the modular file as suggested. Please let me know if there’s anything else I should refine or improve further.

Thanks again for your support!

Cyrilvallez commented on src/transformers/models/telechat2/modular_telechat2.py:
```diff
@@ -0,0 +1,601 @@
+_CONFIG_FOR_DOC = "TeleChat2Config"
+
+
+def dropout_add(x: torch.Tensor, prob: float, training: bool) -> torch.Tensor:
+    """
+    Apply dropout to `x` and add the result to the residual tensor.
+
+    Args:
+        x (`torch.Tensor`): Input tensor.
+        residual (`torch.Tensor`): Residual tensor.
+        prob (`float`): Dropout probability.
+        training (`bool`): Whether in training mode.
+
+    Returns:
+        `torch.Tensor`: Output after dropout and addition.
+    """
+    out = F.dropout(x, p=prob, training=training)
+    return out
```
We don't want to add a new function only for dropout!

Cyrilvallez commented on src/transformers/models/telechat2/modular_telechat2.py:
```diff
@@ -0,0 +1,601 @@
+    def __init__(self, config: TeleChat2Config):
+        super().__init__()
+        hidden_size = config.hidden_size
+        self.gate_proj = nn.Linear(hidden_size, config.ffn_hidden_size, bias=config.mlp_bias)
+        self.up_proj = nn.Linear(hidden_size, config.ffn_hidden_size, bias=config.mlp_bias)
+        self.down_proj = nn.Linear(config.ffn_hidden_size, hidden_size, bias=True)
```
Here instead of redefinition, we want to change the name in the config to match Llama!

Cyrilvallez commented on src/transformers/models/telechat2/modular_telechat2.py:
```diff
@@ -0,0 +1,601 @@
+        self.attention_dropout = config.attention_dropout
+        self.is_causal = True
+
+        self.query = nn.Linear(config.hidden_size, config.num_attention_heads * self.head_dim, bias=False)
+        self.key_value = nn.Linear(config.hidden_size, self.head_dim * config.num_key_value_heads * 2, bias=False)
+        self.dense = nn.Linear(config.num_attention_heads * self.head_dim, config.hidden_size)
```
Similarly here, we want to split the k/v module in 2 separate modules, and rename them to match Llama!

Cyrilvallez reviewed (commented):
Hey! Basically, the easiest (if you do not have many models checkpoints on the hub already?) is to match your config arguments names and model layer names to existing ones in Llama (it is extremely important for everyone to always see the same names so that they know what they are immediately, instead of figuring it out for every model/config). For the config it is straightforward, however model checkpoints will require a conversion script as for e.g. https://github.com/huggingface/transformers/blob/main/src/transformers/models/glm/convert_glm_weights_to_hf.py to change the layer names, and in your case split the k/v layers in 2!

Doing so will greatly simplify the modular as you will be able to inherit everything directly! (I did not review all the file, just the beginning so that you can understand my point!)

Cyrilvallez commented on src/transformers/models/telechat2/modular_telechat2.py:
```diff
@@ -0,0 +1,601 @@
+        del self.norm
+        self.word_embeddings = nn.Embedding(config.vocab_size, config.hidden_size, self.padding_idx)
+
+        self.h = nn.ModuleList([TeleChat2DecoderLayer(config, i) for i in range(config.num_hidden_layers)])
+        self.ln_f = TeleChat2RMSNorm(config.hidden_size, eps=config.layer_norm_epsilon)
```
For example here we want to be much more explicit in the names, see my previous review!

shunxing12345 committed:
add weight convert

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 commented:
Hi @Cyrilvallez 

Thanks for your guidance and suggestions! I’ve addressed the issues and completed the necessary updates as per your feedback. Please let me know if there’s anything else I should refine. 😊

Best regards!

shunxing12345 commented:
hi @Cyrilvallez 
Just a gentle follow-up regarding the PR. When you have a moment, could you kindly review it? Please let me know if there's anything further you'd like me to adjust or refine.

Thank you so much for your time!

Best regards

Cyrilvallez commented on src/transformers/models/telechat2/modular_telechat2.py:
```diff
@@ -0,0 +1,142 @@
+        self.gate_proj = nn.Linear(hidden_size, config.intermediate_size, bias=False)
+        self.up_proj = nn.Linear(hidden_size, config.intermediate_size, bias=False)
+        self.down_proj = nn.Linear(config.intermediate_size, hidden_size, bias=True)
+        self.hidden_dropout = config.hidden_dropout
+
+    def forward(self, hidden_states):
+        intermediate_output = self.down_proj(F.silu(self.gate_proj(hidden_states)) * self.up_proj(hidden_states))
+        output = F.dropout(intermediate_output, p=self.hidden_dropout, training=self.training)
```
Here, if the dropout is always 0 as is the default in the config, let's remove it and just use Llama's forward directly (by adding `config.hidden_act = "silu"` in the config)

shunxing12345 replied:
Thank you for the suggestion! However, the hidden_dropout parameter is indeed always set to 0.0 during inference, but during training, it needs to be set to 0.1 to ensure proper regularization. Removing it entirely would affect training performance. Let me know if you have further thoughts!

Cyrilvallez replied:
For sure! But what I meant is that even during training it seems to be set to 0 all the time. At least according to your models checkpoints I looked up on the hub. So if that's really the case we would prefer to remove it entirely 🤗

shunxing12345 replied:
Got it! Please let me know if there's anything else that needs to be adjusted!

Cyrilvallez commented on src/transformers/models/telechat2/modular_telechat2.py:
```diff
@@ -0,0 +1,142 @@
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        input_shape = hidden_states.shape[:-1]
+        hidden_shape = (*input_shape, -1, self.head_dim)
+
+        query_states = self.q_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+        key_states = self.k_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+        value_states = self.v_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+
+        cos, sin = position_embeddings
+        query_states, key_states = apply_rotary_pos_emb(query_states, key_states, cos, sin)
+
+        if past_key_value is not None:
+            # sin and cos are specific to RoPE models; cache_position needed for the static cache
+            cache_kwargs = {"sin": sin, "cos": cos, "cache_position": cache_position}
+            key_states, value_states = past_key_value.update(key_states, value_states, self.layer_idx, cache_kwargs)
+
+        attention_interface: Callable = eager_attention_forward
+        if self.config._attn_implementation != "eager":
+            if self.config._attn_implementation == "sdpa" and kwargs.get("output_attentions", False):
+                logger.warning_once(
+                    "`torch.nn.functional.scaled_dot_product_attention` does not support `output_attentions=True`. Falling back to "
+                    'eager attention. This warning can be removed using the argument `attn_implementation="eager"` when loading the model.'
+                )
+            else:
+                attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
+
+        attn_output, attn_weights = attention_interface(
+            self,
+            query_states,
+            key_states,
+            value_states,
+            attention_mask,
+            dropout=0.0 if not self.training else self.attention_dropout,
+            scaling=self.scaling,
+            **kwargs,
+        )
+
+        attn_output = attn_output.reshape(*input_shape, -1).contiguous()
+        attn_output = self.o_proj(attn_output)
+        return attn_output, attn_weights
```
This should be removed as it is similar to Llama (it will be picked up automatically by modular)

Cyrilvallez commented on tests/models/telechat2/test_modeling_telechat2.py:
```diff
@@ -0,0 +1,616 @@
+# coding=utf-8
+# Copyright 2024 The TeleChat team, Alibaba Group and The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2025 The TeleChat team and The HuggingFace Inc. team. All rights reserved.
```

Cyrilvallez reviewed (commented):
Hey! Sorry for the delay! Here is a new round of review! 🤗

shunxing12345 committed:
fix doc & del duplicate func

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
inherit llama in test

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
Merge branch 'main' into fix_conflicts

shunxing12345 commented:
Hi @Cyrilvallez ,

Thanks for the review! I have addressed all the comments, and the changes have been updated. Please let me know if there is anything else to improve.

Cheers! 😊

shunxing12345 committed:
remove hidden dropout

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
Merge branch 'fix_conflicts' of https://github.com/shunxing12345/transformers into fix_conflicts

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 commented:
hi @Cyrilvallez ,
Thank you so much for your thoughtful guidance! I have completed all the requested changes and carefully reviewed the updates.

shunxing12345 commented:
Hi @Cyrilvallez,

I hope this message finds you well. I just wanted to kindly follow up regarding the updates I’ve made based on your valuable feedback. Please let me know if there’s anything else I can refine or improve.

Thank you so much for your time and guidance!

Best regards,

Cyrilvallez commented on src/transformers/models/telechat2/modular_telechat2.py:
```diff
@@ -0,0 +1,75 @@
+class TeleChat2Attention(LlamaAttention):
+    def __init__(self, config: TeleChat2Config, layer_idx: int):
+        super().__init__(config, layer_idx)
+        self.q_proj = nn.Linear(
+            config.hidden_size, config.num_attention_heads * self.head_dim, bias=config.attention_bias
+        )
+        self.k_proj = nn.Linear(
+            config.hidden_size, config.num_key_value_heads * self.head_dim, bias=config.attention_bias
+        )
+        self.v_proj = nn.Linear(
+            config.hidden_size, config.num_key_value_heads * self.head_dim, bias=config.attention_bias
+        )
```
This is similar to Llama and is not needed. See previous comment to maybe hardcode bias value though

Cyrilvallez reviewed (commented):
Hey! Let's rework the Config to make it simpler and use modular! 🤗

shunxing12345 committed:
inherited llama config

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
Merge branch 'fix_conflicts' of https://github.com/shunxing12345/transformers into fix_conflicts

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
Merge branch 'main' into fix_conflicts

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 commented:
HI, @Cyrilvallez 
The config has been updated as requested! However, it seems that some errors in the autotest are related to HTTP issues. Let me know if you have any insights on this. 😊

shunxing12345 committed:
Merge branch 'main' into fix_conflicts

shunxing12345 committed:
Merge branch 'main' into fix_conflicts

shunxing12345 committed:
fix

Signed-off-by: shunxing12345 <747777271@qq.com>

shunxing12345 commented:
@Cyrilvallez , Hello!
All the automated tests for this PR have passed, and we’ve addressed the previously mentioned HTTP issues. Could you please take a moment to review it again? If there’s any feedback or suggestions, let me know. Thank you so much!

shunxing12345 commented:
Hi @Cyrilvallez ,

I hope you're doing well! Just wanted to follow up on this PR—since all automated tests have passed and the previous HTTP issues have been addressed, could you kindly take a moment to review it again? Let me know if there's anything else that needs improvement.

Looking forward to your feedback. Thanks a lot! 😊

ArthurZucker commented:
@Cyrilvallez had a conference this week let me have a final look!

shunxing12345 commented:
Hi @ArthurZucker ,

Thanks for the update! I appreciate you taking the time to review it. Let me know if you need any clarifications or further modifications. Looking forward to your feedback!

Cyrilvallez reviewed (commented):
Hey!! So sorry for the late review! Here are a few comments to still address but looks nice! 🤗

ArthurZucker commented on tests/models/telechat2/test_modeling_telechat2.py:
```diff
@@ -0,0 +1,601 @@
+    )
+
+
+class TeleChat2ModelTester(LlamaModelTester):
```
Here let's do like `test_modeling_gemma2` where we just specify the class and inherit the rest

ArthurZucker commented on tests/models/telechat2/test_modeling_telechat2.py:
```diff
@@ -0,0 +1,601 @@
+
+
+@require_torch
+class TeleChat2IntegrationTest(unittest.TestCase):
```
nice! very important thanks for adding!

ArthurZucker reviewed (approved):
LGTM! thanks for the contribution! One nit about test and lets merge!

shunxing12345 commented:
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
